### PR TITLE
Support other subclasses of `EntityManagerInterface` too

### PR DESCRIPTION
**Pull request by NoelLH**
## Description
Support any `EntityManagerInterface` in new `generateId()` methods, and also in existing `generate()`s.

Also require PHP 7.4+ and expand use of type constraints.

I suggest this targets a new major release.

## Motivation and context
Detailed in https://github.com/thebiggive/matchbot/pull/338

## How has this been tested?
Unit tests run and added to. Fork used in https://github.com/thebiggive/matchbot/pull/338 both with real local calls and unit tests.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

I believe this is backwards incompatible only insofar as PHP version support, so I have not made any documentation changes or identified any that are needed.

## PR checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING.md** document.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: src/UuidGenerator.php [hunk 1]
```diff
@@ -27,14 +27,32 @@ class UuidGenerator extends AbstractIdGenerator
...
+     *
+     * @param EntityManagerInterface $em
      * @param Entity $entity
      *
      * @return UuidInterface
      *
      * @throws Exception
      */
-    public function generate(EntityManager $em, $entity)
+    public function generateId(EntityManagerInterface $em, $entity): UuidInterface
```
Comment (ramsey) [inline src/UuidGenerator.php:55]: This introduces a BC break, right? If anyone extends `UuidGenerator` and overrides this method, making this change will result in an error like:

> Fatal error: Declaration of Foo::generate(EntityManager $em, $entity) must be compatible with Ramsey\Uuid\Doctrine\UuidGenerator::generate(EntityManagerInterface $em, $entity): UuidInterface

**Review by ramsey — COMMENTED** [review]
In general, I want to accept these changes, but I think it will require a major version bump, due to the BC break.

File: src/UuidGenerator.php (hunk 1 above)
Comment (NoelLH) [reply src/UuidGenerator.php:55]: @ramsey Yes, sorry I wasn't clear. This is what I was trying to get at with:

> I suggest this targets a new major release.

I would also expect the bump in minimum PHP version to necessitate a new major version as that's also a BC break, hence my thinking that the two changes could go hand in hand.

**Review by ramsey — APPROVED** [review]

**Comment by ramsey** [comment]
Thank you for contributing! 🎉

I committed this in 53955b785a0cac7128bdc6ecdd2e97d3e3391e63

File: src/UuidGenerator.php (hunk 1 above)
Comment (bdsl) [reply src/UuidGenerator.php:55]: A bump in minimum PHP version shouldn't be considered a BC break IMHO. The reasoning is explained under "_Why dropping PHP support in a minor version is not a BC break_" at https://www.doctrine-project.org/2017/07/25/php-7.1-requirement-and-composer.html

**Pull request closed without merging**